File: internal/message/normalize.go
package message

import (
	"strings"
	"unicode"
	"unicode/utf8"

	"github.com/JeanpierreSolis15/commitron/internal/config"
)

func (p Parsed) Canonical() string {
	var b strings.Builder
	b.WriteString(strings.ToLower(p.Type))
	if p.Scope != "" {
		b.WriteString("(" + p.Scope + ")")
	}
	if p.Breaking {
		b.WriteString("!")
	}
	b.WriteString(": ")
	b.WriteString(strings.TrimRight(strings.TrimSpace(p.Description), "."))
	return b.String()
}

func (p Parsed) Render(cfg config.Config) string {
	subject := p.Canonical()
	if p.Body == "" {
		return subject
	}
	return subject + "\n\n" + wrapBody(p.Body, cfg.BodyMaxLineLength)
}

func wrapBody(body string, max int) string {
	if max <= 0 {
		return body
	}
	var out []string
	for _, line := range strings.Split(body, "\n") {
		out = append(out, wrapLine(line, max)...)
	}
	return strings.Join(out, "\n")
}

func wrapLine(line string, max int) []string {
	if utf8.RuneCountInString(line) <= max || strings.TrimSpace(line) == "" {
		return []string{line}
	}

	trimmed := strings.TrimLeft(line, " \t")
	indent := line[:len(line)-len(trimmed)]
	continuation := indent
	if strings.HasPrefix(trimmed, "- ") || strings.HasPrefix(trimmed, "* ") {
		continuation = indent + "  "
	}

	words := strings.Fields(trimmed)
	if len(words) == 0 {
		return []string{line}
	}

	var lines []string
	current := indent + words[0]
	for _, word := range words[1:] {
		if utf8.RuneCountInString(current)+1+utf8.RuneCountInString(word) <= max {
			current += " " + word
			continue
		}
		lines = append(lines, current)
		current = continuation + word
	}
	return append(lines, current)
}

func violatesLowerCase(description string) bool {
	if description == "" {
		return false
	}
	runes := []rune(description)
	if !unicode.IsUpper(runes[0]) {
		return false
	}
	word := description
	if i := strings.IndexAny(description, " \t"); i > 0 {
		word = description[:i]
	}
	for _, r := range []rune(word)[1:] {
		if unicode.IsUpper(r) {
			return false
		}
	}
	return true
}

func hasUpper(s string) bool {
	for _, r := range s {
		if unicode.IsUpper(r) {
			return true
		}
	}
	return false
}

File: src/normalize.ts
import type { Config } from "./config";
import type { Parsed } from "./message";

const upperRe = /\p{Lu}/u;

export function runeCount(s: string): number {
  return Array.from(s).length;
}

export function canonical(p: Parsed): string {
  let subject = p.type.toLowerCase();
  if (p.scope !== "") {
    subject += `(${p.scope})`;
  }
  if (p.breaking) {
    subject += "!";
  }
  return `${subject}: ${p.description.trim().replace(/\.+$/, "")}`;
}

export function render(p: Parsed, cfg: Config): string {
  const subject = canonical(p);
  if (p.body === "") {
    return subject;
  }
  return `${subject}\n\n${wrapBody(p.body, cfg.bodyMaxLineLength)}`;
}

export function wrapBody(body: string, max: number): string {
  if (max <= 0) {
    return body;
  }
  return body
    .split("\n")
    .flatMap((line) => wrapLine(line, max))
    .join("\n");
}

function wrapLine(line: string, max: number): string[] {
  if (runeCount(line) <= max || line.trim() === "") {
    return [line];
  }

  const trimmed = line.replace(/^[ \t]+/, "");
  const indent = line.slice(0, line.length - trimmed.length);
  const continuation =
    trimmed.startsWith("- ") || trimmed.startsWith("* ") ? `${indent}  ` : indent;

  const words = trimmed.split(/\s+/).filter((word) => word !== "");
  if (words.length === 0) {
    return [line];
  }

  const lines: string[] = [];
  let current = indent + words[0];
  for (const word of words.slice(1)) {
    if (runeCount(current) + 1 + runeCount(word) <= max) {
      current += ` ${word}`;
      continue;
    }
    lines.push(current);
    current = continuation + word;
  }
  lines.push(current);
  return lines;
}

export function violatesLowerCase(description: string): boolean {
  if (description === "") {
    return false;
  }
  const chars = Array.from(description);
  if (!upperRe.test(chars[0]!)) {
    return false;
  }
  const space = description.search(/[ \t]/);
  const word = space > 0 ? description.slice(0, space) : description;
  return !Array.from(word)
    .slice(1)
    .some((ch) => upperRe.test(ch));
}

export function hasUpper(s: string): boolean {
  return Array.from(s).some((ch) => upperRe.test(ch));
}
